### ISSUE TITLE
fix for dlopen "symbol not found in flat namespace" error

### ISSUE THREAD

**Issue by dperlman** [invoke-ai/InvokeAI#580]
A number of people, including myself, have encountered the error "symbol not found in flat namespace" when we get to the step of "python scripts/preload_models.py". I wrestled with this for a while after explicitly following the M1 Mac install instructions in INSTALL_MAC.md. I eventually found that the problem was that protobuf had been installed with homebrew. 

Fix: 
In my case I first had to run
brew uninstall protobuf
then install protobuf with
conda install protobuf

But the appropriate bug fix overall would be:
-remove protobuf from the brew install instruction in INSTALL_MAC.md
-add protobuf to environment-mac.yaml

I hope this is helpful, I was not able to find any direct recommendations in extensive searching up to now.

**Comment by iamstickfigure** [comment]
Yep, I verified this works. Definitely agree that the instructions and environment should be updated

**Comment by Vargol** [comment]
which version of protobuf did you install, I'm getting the error in 3.20.x and 3.19.x from pip, 4.x is incompatible with a stack of other dependencies

**Comment by Vargol** [comment]
ha looks like 3.19.4 works, but now hugginface is down

**Comment by mefengl** [comment]
just
`brew uninstall protobuf`
and then 
`conda install protobuf`

**Comment by dperlman** [comment]
Yes, to be clear, my experience is that it *only* works if I install protobuf with conda, neither brew nor pip works. 

My best guess is that "The problem was caused by mixing objects that compiled with libc++ and object that compiled with libstdc++" as described here: 

https://stackoverflow.com/questions/35006614/what-does-symbol-not-found-expected-in-flat-namespace-actually-mean

And in one of the answers there is a link to a Linux Journal article that goes into even more technical detail. However for our purposes I think it's enough to just use conda and not worry about it at this level of detail.

**Comment by Vargol** [comment]
you have to install that every specific version with will or it tries to install a 3.20 release where that symbol was deprecated or a 4.x release which is incompatible with 4 or 5 torch dependancies.

I am not using conda or brew at all.

**Comment by ipenywis** [comment]
I'm using _conda_  and _brew_ but still getting the `symbol not found error`. I tried to install protobuf using _conda_ but no luck

**Comment by cprivitere** [comment]
```sh
brew uninstall protobuf
conda install protobuf
```
Worked here on Ventura.

**Comment by tildebyte** [comment]
Can someone who is using conda (who is seeing this issue) please try (in the virtual environment, of course)

`conda uninstall protobuf`
`pip install protobuf==3.19.4 -U`

and then see if the error persists?

**Comment by ipenywis** [comment]
It works for me now! 
I had to downgrade to _protobuf_ `3.20.1` (_conda_ was installing `3.20.2`)
Thanks to this reply https://github.com/protocolbuffers/protobuf/issues/10571#issuecomment-1247969637

**Comment by ParityError** [comment]
> It works for me now! (M1 Max) I had to downgrade to _protobuf_ `3.20.1` (_conda_ was installing `3.20.2`)
> 
> `conda uninstall protobuf` `conda install protobuf=3.20.1`
> 
> Thanks to this reply [protocolbuffers/protobuf#10571 (comment)](https://github.com/protocolbuffers/protobuf/issues/10571#issuecomment-1247969637)

Using the command `conda uninstall protobuf` followed by `conda install protobuf=3.20.1` somehow borked the `ldm` environment completely in my case. I had to end up doing:

```
conda deactivate
pyenv virtualenv-delete anaconda3-2022.05/envs/ldm
PIP_EXISTS_ACTION=w CONDA_SUBDIR=osx-arm64 conda env create -f environment-mac.yaml
conda activate ldm
conda install protobuf=3.20.1
```

Apart from the deprecated warnings the `python scripts/preload_models.py` command worked as expected.

**Comment by Skrelan** [comment]
Running on MacOS Monterey (12.3.1 ) and M1 chipset, this fix works for me

**Comment by hexsecs** [comment]
> just `brew uninstall protobuf` and then `conda install protobuf`

This worked for me on MacOS Monterey 12.5

**Comment by visualinventor** [comment]
Came here looking for answers for the protobuf namespace issue this is a good solve. Thanks @dperlman

**Comment by rozek** [comment]
Thank you very much for all your infos - I've combined them into a pull request (#671)

**Comment by psychedelicious** [comment]
The fixes in #671 were cherry-picked into development (and now main).
